File: src/main/java/model/moves/BaseMove.java
package model.moves;

import model.GameData;
import model.pieces.Piece;
import model.util.Position;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

/**
 * A move that moves one piece to a new square (and eats the piece at that square if it exists)
 */
public class BaseMove extends Move {
    @Nullable
    private Piece eatenPiece;

    public BaseMove(Position debut, @NotNull Position fin) {
        super(debut, fin);
    }

    @Override
    void applyToGame(GameData data) {
        piece = data.getBoard().removePiece(start); //Remove the piece from the starting position
        eatenPiece = data.getBoard().add(end, piece); //Place the piece at the ending position and get the piece that was replaced

        if (eatenPiece != null) {
            data.getEatenPieces(eatenPiece.getColour()).push(eatenPiece); //Add the eaten to the stack in the game data
        }
    }

    @Override
    void undoToGame(GameData data) {
        data.getBoard().movePiece(start, piece); //Move the piece to its starting position

        //If a piece was eaten remove it from the stack and add it to the board
        if (eatenPiece != null) {
            data.getEatenPieces(eatenPiece.getColour()).pop();
            data.getBoard().add(end, eatenPiece);
        }
    }

    /**
     * @return 0 if no piece was eating (moving a piece does not change the board value). If a piece was eaten the value is - the piece's value
     */
    @Override
    public int getValue() {
        return eatenPiece == null ? 0 : -eatenPiece.getSignedValue();
    }
}

File: engine/src/main/java/engine/moves/BaseMove.kt
package engine.moves

import engine.GameData
import engine.pieces.Piece
import engine.util.Position

/**
 * A move that moves one pieceMap to a new square (and eats the pieceMap at that square if it exists)
 */
internal open class BaseMove(start: Position, destination: Position) : Move(start, destination) {
    protected var eatenPiece: Piece? = null

    /**
     * @return 0 if no pieceMap was eating (moving a pieceMap does not change the pieceMap value). If a pieceMap was eaten the value is - the pieceMap's value
     */
    override val value: Int
        get() = -(eatenPiece?.signedValue ?: 0)


    override fun applyToGame(data: GameData) {
        piece = data.pieceMap.removePiece(start) //Remove the pieceMap from the starting position
        eatenPiece = data.pieceMap.put(end, piece) //Place the pieceMap at the ending position and get the pieceMap that was replaced

        eatenPiece?.let {
            data.getEatenPieces(it.colour).push(it) //Add the eaten to the stack in the game data
        }
    }

    override fun undoToGame(data: GameData) {
        data.pieceMap.movePiece(start, piece) //Move the pieceMap to its starting position

        //If a pieceMap was eaten remove it from the stack and put it to the pieceMap
        eatenPiece?.let {
            data.getEatenPieces(it.colour).pop()
            data.pieceMap.put(end, it)
        }
    }
}
